packages/web3-providers/src/Approval/index.ts

```diff
@@ -0,0 +1,176 @@
+            const fromBlock = approvalListState.tokenState[account]?.[chainId]?.fromBlock ?? 0
+            const toBlock = await web3.eth.getBlockNumber()
+            const logs = await web3.eth.getPastLogs({
+                topics: [TOKEN_APPROVAL_TOPIC, web3.eth.abi.encodeParameter('address', account)],
+                fromBlock,
+                toBlock,
+            })
+            const maskDappContractInfoList = getAllMaskDappContractInfo(chainId, 'token')
+
+            const tokenSpenderRecords = parseLogs(web3, logs)
+
+            mapKeys(tokenSpenderRecords, (spender, spenderAddress) =>
+                mapKeys(spender, (token, tokenAddress) => {
+                    const latestTx = maxBy(token, (x) => x.blockNumber)
+                    if (!latestTx) return
+                    const amount = new BigNumber(isZeroAddress(latestTx.data) ? 0 : latestTx.data)
+                    approvalListState.updateTokenState(
+                        account,
+                        spenderAddress,
+                        tokenAddress,
+                        chainId,
+                        toBlock,
+                        latestTx.blockNumber,
+                        amount,
+                    )
+                }),
+            )
+
+            const spenderList = approvalListState.tokenState[account]?.[chainId]?.spenderList
+
+            if (!spenderList) return EMPTY_LIST
+
+            return Object.keys(spenderList)
+                .map((spender) => {
+                    return Object.keys(spenderList[spender]).map((address) => {
+                        const maskDappContractInfo = maskDappContractInfoList.find((y) =>
+                            isSameAddress(y.address, spender),
+                        )
+
+                        return {
+                            tokenInfo: { address, name: '', symbol: '' },
+                            address: spender,
+                            name: maskDappContractInfo?.name,
+                            logo: maskDappContractInfo?.logo,
+                            rawAmount: spenderList[spender][address].amount.toNumber(),
+                            transactionBlockNumber: spenderList[spender][address].transactionBlockNumber,
+                            isMaskDapp: Boolean(maskDappContractInfo),
+                        }
+                    })
+                })
+                .flat()
+                .filter((x) => isGreaterThan(x.rawAmount, 0))
+                .sort((a, b) => {
+                    if (a.isMaskDapp && !b.isMaskDapp) return -1
+                    if (!a.isMaskDapp && b.isMaskDapp) return 1
+                    return Number(b.transactionBlockNumber) - Number(a.transactionBlockNumber)
+                }) as Array<FungibleTokenSpender<ChainId, SchemaType>>
+        } catch (error) {
+            console.log(error)
+            return EMPTY_LIST
+        }
+    }
+
+    async getNonFungibleTokenSpenders(chainId: ChainId, account: string) {
```

Comment:
This method is sharing a bit part with `getNonFungibleTokenSpenders`. Merge them?

Reply:
done